src/core/MIGRATION_EXAMPLES.md

```diff
@@ -450,6 +451,140 @@ class Plugin {
 }
 ```
 
+### Migrating Hapi "pre" handlers
+
+In the Legacy Platform, routes could provide a "pre" option in their config to
+register a function that should be run prior to the route handler. These
+"pre" handlers allow routes to share some business logic that may do some
+pre-work or validation. In Kibana, these are often used for license checks.
+
+The Kibana Platform's HTTP interface does not provide this functionality,
+however it is simple enough to port over using a higher-order function that can
+wrap the route handler.
+
+#### Simple example
+
+In this simple example, a pre-handler is used to either abort the request with
+an error or continue as normal. This is a simple "gate-keeping" pattern.
+
+```ts
+// Legacy pre-handler
+const licensePreRouting = (request) => {
+  const licenseInfo = getMyPluginLicenseInfo(request.server.plugins.xpack_main);
+  if (!licenseInfo.isHighEnough()) {
+    throw Boom.forbidden(`You don't have the right license for MyPlugin!`);
+  }
+}
+
+server.route({
+  method: 'GET',
+  path: '/api/my-plugin/do-something',
+  config: {
+    pre: [{ method: licensePreRouting }]
+  },
+  handler: (req) => {
+    return doSomethingInteresting();
+  }
+})
+```
+
+In the Kibana Platform, the same functionality can be acheived by creating a
+function that takes a route handler (or factory for a route handler) as an
+argument and either invokes it in the successful case or returns an error
+response in the failure case.
+
+We'll call this a "high-order handler" similar to the "high-order component"
+pattern common in the React ecosystem.
+
+```ts
+// New Platform high-order handler
+const checkLicense = <P, Q, B>(
```

Comment:
That would be a pain for the plugins to declare it manually. Should we provide a type for the wrapper?
```typescript
// platform code
type RequestHandlerWrapper = <P, Q, B>(
  handler: RequestHandler<P, Q, B, RouteMethod>
) => RequestHandler<P, Q, B, RouteMethod>;
// plugin code
const checkLicense: RequestHandlerWrapper = handler => ....
```

Reply:
This type is essentially the same as:
```ts
type FunctionWrapper = <T>(x: T) => T;
```

Maybe we could just put this in kbn/utility-types?

Reply:
Going to merge this now, but we can still decide on this and make an update to the doc.

Reply:
the problem with this that it doesn't constrain `T` type.